contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,846 @@
+
+### StatefulSet Revision History
+The StatefulSet controller will use labeled, versioned PodTemplates to keep a 
+history of updates performed on a StatefulSet. The number of stored PodTemplates 
+is considered to be the size of the StatefulSet's revision history. The 
+maximum size of a StatefulSet's revision history is two (these are the current 
+and target PodTemplates) plus the history limit (represented by its 
+`.Spec.RevisionHistoryLimit`).
+
+#### PodTemplate Creation
+When the `.Spec.Template` of a StatefulSet is [updated](#template-updates), the 
+StatefulSet controller will create a new PodTemplate to represent the new 
+revision of the StatefulSet. When the controller creates a PodTemplate for a 
+StatefulSet, it will do the following.
+
+1. The controller will set the PodTemplate's `.PodTemplateSpec` field to the 
+StatefulSet's `.Spec.Template` field.
+1. The controller will create a ControllerRef object in the PodTemplate's 
+`.OwnerReferences` list to mediate selector overlapping.
+1. The controller will label the PodTemplate with a with a label matching the 
+StatefulSet's `.Spec.Selector` to allow for the selection of the PodTemplate.
+1. The controller will label the PodTemplate's PodTemplateSpec with a 
+`StaefulSetTemplateGenerationLabel` set to the StatefulSet's 
+`.Spec.TemplateGeneration`.
+1. The controller will set the Name of the PodTemplate to a concatenation of the 
+`.Name` of the StatefulSet and the `.Spec.TemplateGeneration`.
+1. The controller will then create the PodTemplate.
+
+#### PodTemplate Deletion
+When the `StatefulSet` controller deletes a PodTemplate in the revision 
+history of a StatefulSet it will do the following.
+
+1. If the PodTemplate's ControllerRef does not match the StatefulSet, the 
+controller will not delete the PodTemplate. In this way, we prevent selector 
+overlap from causing the deletion of PodTemplates that are part of another 
+object's revision history. In practice, these PodTemplates will be filtered out
+prior to history maintenance.
+1. If the PodTemplate's ControllerRef matches the StatefulSet, the 
+StatefulSet controller will delete the PodTemplate.
+
+#### History Reconstruction
+In order to reconstruct the history of revisions to a StatefulSet, the 
+StatefulSet controller will do the following.
+
+1. If the StatefulSet's `.Spec.TemplateGeneration` is nil, the StatefulSet 
+has never been updated, and its history has never been initialized. This is 
+the state the object will be in when a cluster is first upgraded from a version
+that does not support StatefulSet update to a version that does. In this case,
+the controller will not enforce PodTemplate revisions. When creating Pods, 
+it will always use the StatefulSet's `.Spec.Template`. Otherwise, the controller 
+will continue as below.
+1. The controller will select all PodTemplates with a 
+`StatefulSetPodTemplateLabel` matching the `.Name` field of the StatefulSet.
```

Comment:
That needs to be update to reflect selection based on the Selector. The initial thought was that you could use the .Name of the StatefulSet as a unique selector, but as @enisoc  pointed out, we might as well just use the existing selector.